storages/backends/azure_storage.py

```diff
@@ -1,42 +1,123 @@
+
+from azure.common import AzureMissingResourceHttpError
+from azure.storage.blob import ContentSettings
+import base64
+from django.utils.six.moves import urllib
 
 
 def clean_name(name):
     return os.path.normpath(name).replace("\\", "/")
 
 
+def pad_left(n, width, pad="0"):
+    return ((pad * width) + str(n))[-width:]
+
+
+@deconstructible
+class AzureStorageFile(File):
+
+    def __init__(self, name, mode, storage):
+        self._name = name
+        self._mode = mode
+        self._storage = storage
+        self._is_dirty = False
+        self._file = None
+        if 'w' in self._mode:
+            self._storage.connection._put_blob(self._storage.azure_container, self._name, None)
+        self._write_counter = 0
+        self._block_list = list()
+        self._last_commit_pos = 0
+
+    def _get_file(self):
+        if self._file is None:
+            self._file = SpooledTemporaryFile(
+                max_size=self._storage.max_memory_size,
+                suffix=".AzureBoto3StorageFile",
+                dir=setting("FILE_UPLOAD_TEMP_DIR", None)
+            )
+            if 'r' in self._mode:
+                self._is_dirty = False
+                # I set max connection to 1 since spooledtempfile is not seekable which is required if we use
+                # max_conection > 1
+                self._storage.connection.get_blob_to_stream(container_name=self._storage.azure_container,
+                                                            blob_name=self._name, stream=self._file,
+                                                            max_connections=1)
+
+                self._file.seek(0)
+        return self._file
+
+    file = property(_get_file)
+
+    def read(self, *args, **kwargs):
+        if 'r' not in self._mode:
+            raise AttributeError("File was not opened in read mode.")
+        return super(AzureStorageFile, self).read(*args, **kwargs)
+
+    def write(self, content):
+        if 'w' not in self._mode:
+            raise AttributeError("File was not opened in write mode.")
+        self._is_dirty = True
+        ret = super(AzureStorageFile, self).write(force_bytes(content))
+        if self._needs_flush():
+            self._flush_all_buffers()
+        return ret
+
+    def _needs_flush(self, current_pos=None):
+        if not(current_pos):
+            current_pos = self.file.tell()
+        buffer_size = current_pos - self._last_commit_pos
+        ret_val = buffer_size >= self._storage.buffer_size
+        return ret_val
+
+    def _flush_buffer(self):
+        self._write_counter += 1
+        block_id = force_bytes(pad_left("{}{}".format(self._name, self._write_counter), 32), 'utf-8')
+        block_id = base64.urlsafe_b64encode(block_id)
+        block_id = urllib.parse.quote_plus(block_id)
+        self.file.seek(self._last_commit_pos)
+        content = self.file.read(self._storage.buffer_size)
+        self._storage.connection.put_block(self._storage.azure_container, self._name,
```

Comment:
Looks like the azure api errors out if `len(content) == 0`:

```
>>> content = b""
>>> AzureStorage().connection.put_block(cas.azure_container, 'test.tmp',  content, 0)
Traceback (most recent call last):
  File "<console>", line 1, in <module>
  File "C:\Users\H193143\ntl_loader\venv\lib\site-packages\azure\storage\blob\blockblobservice.py", line 191, in put_block
    timeout=timeout
  File "C:\Users\H193143\ntl_loader\venv\lib\site-packages\azure\storage\blob\blockblobservice.py", line 959, in _put_block
    self._perform_request(request)
  File "C:\Users\H193143\ntl_loader\venv\lib\site-packages\azure\storage\storageclient.py", line 280, in _perform_request
    raise ex
  File "C:\Users\H193143\ntl_loader\venv\lib\site-packages\azure\storage\storageclient.py", line 248, in _perform_request
    raise ex
  File "C:\Users\H193143\ntl_loader\venv\lib\site-packages\azure\storage\storageclient.py", line 235, in _perform_request
    _http_error_handler(HTTPError(response.status, response.message, response.headers, response.body))
  File "C:\Users\H193143\ntl_loader\venv\lib\site-packages\azure\storage\_error.py", line 114, in _http_error_handler
    raise AzureHttpError(message, http_error.status)
azure.common.AzureHttpError: The value for one of the HTTP headers is not in the correct format.
<?xml version="1.0" encoding="utf-8"?><Error><Code>InvalidHeaderValue</Code><Message>The value for one of the HTTP headers is not in the correct format.
RequestId:ea7efa45-001e-00c1-39a3-593eed000000
Time:2017-11-09T21:39:46.6472454Z</Message><HeaderName>Content-Length</HeaderName><HeaderValue>0</HeaderValue></Error>

```

I got it to work by just adding this above line 90:

```
if not content:
    return
```